src/Build/ElementLocation/ElementLocation.cs

```diff
@@ -185,192 +176,288 @@ internal static ElementLocation Create(string file)
         /// In AG there are 600 locations that have a file but zero line and column.
         /// In theory yet another derived class could be made for these to save 4 bytes each.
         /// </remarks>
-        internal static ElementLocation Create(string file, int line, int column)
+        internal static ElementLocation Create(string? filePath, int line, int column)
         {
-            if (string.IsNullOrEmpty(file) && line == 0 && column == 0)
+            // Combine line and column values with bitwise OR so we can perform various
+            // checks on both values in a single comparison, reducing the amount of branching
+            // in the code.
+            int combinedValue = line | column;
+
+            if (string.IsNullOrEmpty(filePath) && combinedValue == 0)
             {
-                return EmptyLocation;
+                // When combinedValue is zero, it implies that both line and column are zero.
+                return EmptyElementLocation.Instance;
             }
 
-            if (line <= 65535 && column <= 65535)
+            // When combinedValue is negative, it implies that either line or column were negative.
+            ErrorUtilities.VerifyThrow(combinedValue > -1, "Use zero for unknown");
+
+            // TODO store the last run's value and check if this is for the same file. If so, skip the dictionary lookup (tree walk).
```

Comment:
I wonder if storing the last file in a thread-static variable wouldn't amortize the cost of full lookup/add enough that we could use a conventional data structure with a simple lock.